src/steps/check-lead-activity.ts

```diff
@@ -0,0 +1,150 @@
+    const minutesAgo = stepData.minutes;
+    const withAttributes = stepData.withAttributes || {};
+
+    try {
+      const sinceDate = moment().subtract(minutesAgo, 'minutes').utc().format(moment.defaultFormatUtc);
+      const tokenResponse = await this.client.getActivityPagingToken(sinceDate);
+      const nextPageToken = tokenResponse.nextPageToken;
+
+      const lead = (await this.client.findLeadByEmail(email)).result[0];
+
+      /* Error when lead is not found */
+      if (!lead) {
+        return this.error('Lead %s was not found', [
+          email,
+        ]);
+      }
+
+      const activityTypes = (await this.client.getActivityTypes()).result;
+      const activityType = activityTypes.find(type => type[isNaN(activityTypeIdOrName) ? 'name' : 'id'] == activityTypeIdOrName);
+
+      /* Error when activity is not found */
+      if (!activityType) {
+        return this.error('Activity with ID or Name %s was not found.', [
+          stepData.activityTypeIdOrName,
+        ]);
+      }
+
+      activityTypeIdOrName = activityType.id;
+
+      const activityResponse = await this.client.getActivities(nextPageToken, lead.id, activityTypeIdOrName);
+      const activities = activityResponse.result;
+
+      /* Fail when when the activity supplied is not found in the lead's logs. */
+      if (!activities) {
+        return this.fail('Activity %s was not found for lead %s within the last %d minute(s)', [
+          stepData.activityTypeIdOrName,
+          email,
+          minutesAgo,
+        ]);
+      }
+
+      /* Expected attributes passed to test step. Translate object/map as array for easier comparison with actual attributes */
+      const expectedAttributes = Object.keys(withAttributes).map((key) => { return { name: key, value: withAttributes[key] }; });
+
+      /* Assert Actual vs Expected attributes and pass if at least one activity matches attributes. Otherwise fail */
+      if (expectedAttributes.length > 0) {
+        let validated = false;
+        for (const activity of activities) {
+          const primaryAttribute = this.getPrimaryAttribute(activityTypes, activity);
+          const actualAttributes = activity.attributes;
+          if (primaryAttribute) {
+            actualAttributes.push(primaryAttribute);
+          }
+
+          if (this.hasMatchingAttributes(actualAttributes, expectedAttributes)) {
+            validated = true;
+            break;
+          }
+        }
+
+        if (validated) {
+          return this.pass('Activity %s was found for lead %s within the last %s minute(s). With expected attributes: \n\n', [
+            stepData.activityTypeIdOrName,
+            email,
+            minutesAgo,
+            JSON.stringify(expectedAttributes, null, 2),
+          ]);
+        }
+
+        return this.fail('Expected attributes of activity %s for lead %s within the last %d minute(s) did not match the actual activity attributes. Actual attributes: \n\n', [
```

Comment:
- [ ] nit: Message has gone through so many iterations, its become un-parseable to me.  How about...

```
Found %s activity for lead %s within the last %d minute(s), but none matched the expected attributes. Matching activities include:\n\n%s
```